Sources/PitchSpeller/Data Structures/_Graph.swift

```diff
@@ -0,0 +1,293 @@
+        }
+        
+        init (_ nodes: Pair, withWeight weight: Weight) {
+            self.nodes = nodes
+            self.weight = weight
+        }
+    }
+    
+    struct Path {
+        
+        // MARK: - Instance Properties
+        
+        let nodes: [Node]
+        let weights: [Pair: Weight]
+        
+        // MARK: - Initializers
+        
+        init (_ nodes: [Node], _ weights: [Pair: Weight]) {
+            self.nodes = nodes
+            self.weights = weights
+        }
+    }
+    
+    // MARK: - Instance Properties
+     
+    var nodes: Set<Node>
+    var adjacents: [Pair: Weight]
+    
+    var edges: [Edge] {
+        return adjacents.compactMap { Edge($0.key, withWeight: $0.value) }
+    }
+    
+    // MARK: - Instance Methods
+    
+    mutating func insertNode (_ node: Node) {
+        nodes.insert(node)
+    }
+    
+    mutating func insertEdge (from source: Node, to destination: Node, withWeight weight: Weight) {
+        adjacents[Pair(source, destination)] = weight
+    }
+    
+    mutating func insertEdge (_ keyValue: (Pair, Weight)) {
+        insertEdge(from: keyValue.0.a, to: keyValue.0.b, withWeight: keyValue.1)
+    }
+    
+    mutating func insertEdge(_ pair: Pair, _ weight: Weight) {
+        insertEdge(from: pair.a, to: pair.b, withWeight: weight)
+    }
+
+    mutating func updateEdge(_ pair: Pair, with transform: (Weight) -> Weight) {
+        guard let weight = weight(pair) else { return }
+        insertEdge(pair, transform(weight))
+    }
+    
+    mutating func insertPath (_ path: Path) {
+        path.nodes.forEach { insertNode($0) }
+        path.weights.forEach { insertEdge($0) }
+    }
+    
+    mutating func removeEdge (from source: Node, to destination: Node) {
+        adjacents[Pair(source, destination)] = nil
+    }
+    
+    /// - Returns: Weight of the edge from `source` to `destination` if it exists, else nil
+    func weight (from source: Node, to destination: Node) -> Weight? {
+        return adjacents[Pair(source, destination)]
+    }
+    
+    func weight (_ pair: Pair) -> Weight? {
+        return weight(from: pair.a, to: pair.b)
+    }
+    
+    /// - Returns: Array of nodes adjacent to `source`
+    func neighbors (of source: Node) -> [Node] {
+        return nodes.compactMap {
+            guard let _ = adjacents[Pair(source, $0)] else { return nil }
+            return $0
+        }
+    }
+    
+    /// - Returns: Array of nodes adjacent to `source` out of the supplied set of `nodes`.
+    func neighbors (of source: Node, from nodes: Set<Node>) -> [Node] {
+        return nodes.compactMap {
+            guard let _ = adjacents[Pair(source, $0)] else { return nil }
+            return $0
+        }
+    }
+    
+    func neighbors (of source: Node, from nodes: [Node]) -> [Node] {
+        return neighbors(of: source, from: Set(nodes))
+    }
+    
+    /// - Returns: Array of edges emanating from `source`
+    func edges (from source: Node) -> [Edge] {
```

Comment:
Would  a `Set<Edge>` be a better model for this (i.e., does order matter)?

Reply:
Yes. I think as soon as the moment arises where this design is more appropriate (e.g. in a test or other comparison operation/ perhaps even a subtracting operation) we should implement.

Reply:
👍